apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -2100,3 +2223,76 @@ abort_health_checks_for_channel(Data0, ChannelId) ->
         hc_workers = HCWorkers,
         hc_pending_callers = Pending
     }.
+
+collect_and_handle_channel_operations_connected(Op, Data0) ->
+    From = undefined,
+    Ops = collect_and_compress_channel_operations([Op]),
+    lists:foldl(
+        fun
+            (#add_channel{channel_id = ChannelId, config = Config}, {AccActions, AccData}) ->
+                {Actions, Data} = handle_add_channel(From, AccData, ChannelId, Config),
+                {Actions ++ AccActions, Data};
+            (#remove_channel{channel_id = ChannelId}, {AccActions, AccData}) ->
+                {Actions, Data} = handle_remove_channel(From, ChannelId, AccData),
+                {Actions ++ AccActions, Data}
+        end,
+        {[], Data0},
+        Ops
+    ).
+
+collect_and_handle_channel_operations_not_connected(Op, State, Data0) ->
+    From = undefined,
+    Ops = collect_and_compress_channel_operations([Op]),
+    lists:foldl(
+        fun
+            (#add_channel{channel_id = ChannelId, config = Config}, {AccActions, AccData}) ->
+                {Actions, Data} =
+                    handle_not_connected_add_channel(From, ChannelId, Config, State, AccData),
+                {Actions ++ AccActions, Data};
+            (#remove_channel{channel_id = ChannelId}, {AccActions, AccData}) ->
+                {Actions, Data} = handle_remove_channel(From, ChannelId, AccData),
+                {Actions ++ AccActions, Data}
+        end,
+        {[], Data0},
+        Ops
+    ).
+
+collect_and_compress_channel_operations(Acc) ->
+    MaxRequests = 500,
```

Comment:
It's unlikely to collect 500 messages in mailbox, but if it happens, isn't it better to compact them all as soon as possible?

Reply:
yes, as mentioned in the previous comment, I don't think we'll have that many.  it's just an upper limit.

>  isn't it better to compact them all as soon as possible?

isn't that what this function does?  not sure I understand.

Reply:
I mean if there are more than 500, the remaining ones will be handled in the next chunk, possibly takes even more time.

Reply:
I see.  I think it's important to have an upper limit like this, though, or else we might be stuck collecting requests forever if timing and frequency of requests is just right.